pygfx/objects/_events.py

```diff
@@ -0,0 +1,351 @@
+        an actual time stamp so please only use this for relative time measurements."""
+        return self._time_stamp
+
+    @property
+    def bubbles(self) -> bool:
+        """A boolean value indicating whether or not the event bubbles up through
+        the scene tree."""
+        return self._bubbles
+
+    @property
+    def target(self) -> "EventTarget":
+        """The object onto which the event was dispatched."""
+        return self._target
+
+    @property
+    def current_target(self) -> "EventTarget":
+        """The object that is currently handling the event. During event bubbling
+        this property will be updated whenever an event bubbles up in the hierarchy."""
+        return self._current_target
+
+    @property
+    def cancelled(self) -> bool:
+        """A boolean value indicating whether the event is cancelled."""
+        return self._cancelled
+
+    def stop_propagation(self):
+        """Stops propagation of events further along in the scene tree."""
+        self._bubbles = False
+
+    def cancel(self):
+        """Cancels the event and stops propagation."""
+        self._cancelled = True
+        self.stop_propagation()
+
+    def _retarget(self, target):
+        self._target = target
+        self._update_current_target(target)
+
+    def _update_current_target(self, target):
+        self._current_target = target
+
+    def __getitem__(self, key):
+        """Make the extra kwargs available directly on the event object through
+        bracket syntax."""
+        return self._data[key]
+
+
+class KeyboardEvent(Event):
+    def __init__(self, *args, key, modifiers=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.key = key
+        self.modifiers = modifiers or ()
+
+
+class PointerEvent(Event):
+    def __init__(
+        self,
+        *args,
+        x,
+        y,
+        button=0,
+        buttons=None,
+        modifiers=None,
+        ntouches=0,
+        touches=None,
+        clicks=0,
+        pointer_id=0,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.x = x
+        self.y = y
+        self.button = button
+        self.buttons = buttons or ()
+        self.modifiers = modifiers or ()
+        self.ntouches = ntouches
+        self.touches = touches or {}
+        self.clicks = clicks
+        self.pointer_id = pointer_id
+
+    def copy(self, type, clicks):
+        result = PointerEvent(
+            type=type,
+            x=self.x,
+            y=self.y,
+            button=self.button,
+            buttons=self.buttons,
+            modifiers=self.modifiers,
+            ntouches=self.ntouches,
+            touches=self.touches,
+            clicks=clicks,
+            pointer_id=self.pointer_id,
+            bubbles=self.bubbles,
+            target=self.target,
+            cancelled=self.cancelled,
+            time_stamp=self.time_stamp,
+            # **self._data,
```

Comment:
lets make sure we test `copy()`

Reply:
I've added a simple test. If you have something fancier in mind, please let me know!